jurity/recommenders/interlist_diversity.py

```diff
@@ -0,0 +1,214 @@
+# -*- coding: utf-8 -*-
+# Copyright FMR LLC <opensource@fidelity.com>
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import List, Union, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+from sklearn.metrics.pairwise import cosine_similarity
+import warnings
+
+from jurity.utils import Constants
+
+
+def sample_users(df: pd.DataFrame, user_id_column: str = Constants.user_id,
+                 sample_size: float = None, seed: int = Constants.default_seed) -> pd.DataFrame:
+    """
+    Samples input data frame by selecting a random sample of users.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with a user_id_col column.
+    user_id_column: str
+        User id column name.
+    sample_size: float
+        Proportion of users to randomly sample for evaluation.
+    seed : int, default=Constants.default_seed
+        The seed used to create random state.
+
+    Returns
+    -------
+    Sampled data frame
+    """
+    rng = np.random.RandomState(seed)
+    users = df[user_id_column].unique()
+    users = rng.choice(users, size=int(len(users) * sample_size), replace=False)
+    return df[df[user_id_column].isin(users)]
+
+
+def tocsr(df: pd.DataFrame, user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id):
+    """
+    Transform data frame with user_id and item_id columns to sparse csr matrix.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with (user_id, item_id) in each row.
+    user_id_column: str
+        User id column name.
+    item_id_column: str
+        Item id column name.
+
+    Returns
+    -------
+    Sparse matrix with user-item interactions/recommendations.
+    """
+
+    # Map each user_id to (0, n_users)
+    users = list(df[user_id_column].unique())
+    user_map = dict(zip(users, range(len(users))))
+
+    # Map each user_id to (0, n_items)
+    items = list(df[item_id_column].unique())
+    item_map = dict(zip(items, range(len(items))))
+
+    # Update user_id, item_id values
+    user_ids = df[user_id_column].map(user_map).values
+    item_ids = df[item_id_column].map(item_map).values
+
+    # Convert to csr matrix
+    csr_matrix = sp.coo_matrix((np.ones(len(user_ids)), (user_ids, item_ids))).tocsr()
+
+    return csr_matrix
```

Comment:
I guess you use this only to calculate the distances for the diversity metric and do not need to re-use the `user_map` nor `item_map` later, correct? If this function is to be made reusable, would it make sense to also return the mappings (optionally) and save them to disk?

Reply:
if this moves to utils, I don't see what id_column and item_column str means for a to_csr() function

Reply:
> 

I guess it would make sense to optionally return the mappings, while I am a little not sure if this could be more useful for some other metrics.  I am open to hear more thoughts. @dorukkilitcioglu @skadio

Reply:
Added a util function to get the two mappings from original ids to integer ids, to make the function reusable. See [commit](b4d512f8e708fcd2c4b10121009d4a3bc8550c03). Currently I think we might not need to store these mappings though, given they can be created fast and no other metrics are using the mappings currently.